Fix touchup problem on touchpads

Index: xserver-xorg-input-evdev/src/evdev.c
===================================================================
--- xserver-xorg-input-evdev.orig/src/evdev.c	2011-01-28 10:53:25.000000000 +1100
+++ xserver-xorg-input-evdev/src/evdev.c	2011-01-28 12:04:51.761087000 +1100
@@ -384,7 +384,7 @@
     *num_v = *first_v = 0;
 
     /* convert to relative motion for touchpads */
-    if (pEvdev->abs_queued && (pEvdev->flags & EVDEV_RELATIVE_MODE)) {
+    if (pEvdev->flags & EVDEV_RELATIVE_MODE) {
         if (pEvdev->in_proximity) {
             if (pEvdev->old_vals[0] != -1)
                 pEvdev->delta[REL_X] = pEvdev->vals[0] - pEvdev->old_vals[0];
@@ -397,8 +397,10 @@
         } else {
             pEvdev->old_vals[0] = pEvdev->old_vals[1] = -1;
         }
-        pEvdev->abs_queued = 0;
-        pEvdev->rel_queued = 1;
+	if (pEvdev->abs_queued) {
+	    pEvdev->abs_queued = 0;
+	    pEvdev->rel_queued = 1;
+	}
     }
 
     if (pEvdev->rel_queued) {
