Index: xserver-xorg-input-evdev/src/evdev.h
===================================================================
--- xserver-xorg-input-evdev.orig/src/evdev.h	2008-09-12 15:58:54.000000000 +0300
+++ xserver-xorg-input-evdev/src/evdev.h	2008-09-12 16:00:35.000000000 +0300
@@ -41,9 +41,7 @@
 
 #define EVDEV_MAXBUTTONS 32
 
-#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) >= 3
 #define HAVE_PROPERTIES 1
-#endif
 
 #define LONG_BITS (sizeof(long) * 8)
 #define NBITS(x) (((x) + LONG_BITS - 1) / LONG_BITS)
