Index: xserver-xorg-input-evdev/configure.ac
===================================================================
--- xserver-xorg-input-evdev.orig/configure.ac	2011-01-28 12:20:11.311087001 +1100
+++ xserver-xorg-input-evdev/configure.ac	2011-01-28 12:20:13.601087001 +1100
@@ -65,6 +65,22 @@
 AC_ARG_WITH([sdkdir], [], [sdkdir="$withval"])
 AC_SUBST([sdkdir])
 
+AC_ARG_ENABLE(utouch-grail, AS_HELP_STRING([--enable-grail], [Build with utouch-grail gesture support (default: auto)]), [GRAIL=$enableval], [GRAIL=auto])
+PKG_CHECK_MODULES(GRAIL, utouch-grail, [have_grail=yes], [have_grail=no])
+if test "x$GRAIL" = xauto; then
+        GRAIL="$have_grail"
+fi
+if test "x$GRAIL" = xyes; then
+        if test "x$have_grail" = xno; then
+                AC_MSG_ERROR([Grail build explicitly requested, but required modules not found.])
+        fi
+        AC_DEFINE(USE_GRAIL, 1, Whether to use grail or not)
+        AC_DEFINE(GESTURES, 1, Needed to interpret xorg headers properly)
+else
+        AC_DEFINE(USE_GRAIL, 0, Whether to use grail or not)
+fi
+AM_CONDITIONAL(USE_GRAIL, [test "x$GRAIL" = xyes])
+
 DRIVER_NAME=evdev
 AC_SUBST([DRIVER_NAME])
 
Index: xserver-xorg-input-evdev/src/Makefile.am
===================================================================
--- xserver-xorg-input-evdev.orig/src/Makefile.am	2011-01-28 12:20:11.331087001 +1100
+++ xserver-xorg-input-evdev/src/Makefile.am	2011-01-28 12:20:13.601087001 +1100
@@ -29,11 +29,18 @@
 AM_CPPFLAGS =-I$(top_srcdir)/include
 
 @DRIVER_NAME@_drv_la_LTLIBRARIES = @DRIVER_NAME@_drv.la
-@DRIVER_NAME@_drv_la_LDFLAGS = -module -avoid-version
+@DRIVER_NAME@_drv_la_LDFLAGS = -module -avoid-version 
 @DRIVER_NAME@_drv_ladir = @inputdir@
 
+@DRIVER_NAME@_drv_la_LIBS = @GRAIL_LIBS@
+
+if USE_GRAIL
+GRAIL_SRC=evdev-grail.c evdev-grail.h
+endif
+
 @DRIVER_NAME@_drv_la_SOURCES = @DRIVER_NAME@.c \
                                @DRIVER_NAME@.h \
+			       $(GRAIL_SRC) \
                                emuMB.c \
                                emuWheel.c \
                                draglock.c
Index: xserver-xorg-input-evdev/src/evdev-grail.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ xserver-xorg-input-evdev/src/evdev-grail.c	2011-01-28 12:20:13.601087001 +1100
@@ -0,0 +1,313 @@
+/*
+ * Copyright © 2010 Canonical, Ltd.
+ *
+ * Permission to use, copy, modify, distribute, and sell this software
+ * and its documentation for any purpose is hereby granted without
+ * fee, provided that the above copyright notice appear in all copies
+ * and that both that copyright notice and this permission notice
+ * appear in supporting documentation, and that the name of Red Hat
+ * not be used in advertising or publicity pertaining to distribution
+ * of the software without specific, written prior permission.  Red
+ * Hat makes no representations about the suitability of this software
+ * for any purpose.  It is provided "as is" without express or implied
+ * warranty.
+ *
+ * THE AUTHORS DISCLAIM ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
+ * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
+ * NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY SPECIAL, INDIRECT OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ * Authors:
+ *	Chase Douglas (chase.douglas@canonical.com)
+ */
+
+/* So we can get at the right data in xorg/windowstr.h */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <X11/Xatom.h>
+#include <xf86.h>
+#include <xf86Xinput.h>
+#include <exevents.h>
+#include <windowstr.h>
+#include <xorg/os.h>
+
+#include <evdev-properties.h>
+#include "evdev.h"
+#include "evdev-grail.h"
+#include "gestureproto.h"
+
+/* Copied from evdev.c, will be in evdev.h when xserver 1.9 is merged */
+#define EVDEV_RELATIVE_EVENTS   (1 << 2)
+#define EVDEV_TOUCHPAD          (1 << 4)
+
+static WindowPtr CommonAncestor(WindowPtr a, WindowPtr b)
+{
+    WindowPtr c;
+
+    if (a == b)
+        return a;
+
+    for (b = b; b; b = b->parent)
+        for (c = a; c; c = c->parent)
+            if (c == b)
+                return b;
+
+    return NullWindow;
+}
+
+static WindowPtr GetWindowForGestures(struct grail *grail,
+                                      const struct grail_coord *contacts,
+                                      int num_contacts)
+{
+    WindowPtr window = NULL;
+    int i;
+
+    for (i = 0; i < num_contacts; i++)
+    {
+        float screen_x = contacts[i].x;
+        float screen_y = contacts[i].y;
+        WindowPtr this_window;
+
+        this_window = xf86CoordinatesToWindow(screen_x, screen_y, 0);
+        if (!this_window)
+            return NullWindow;
+
+        if (!window)
+            window = this_window;
+        else
+            window = CommonAncestor(window, this_window);
+    }
+
+    return window;
+}
+
+static int GetClients(struct grail *grail,
+		      struct grail_client_info *clients, int max_clients,
+		      const struct grail_coord *contacts, int num_contacts,
+		      const grail_mask_t *types, int type_bytes)
+{
+    WindowPtr child_window;
+    WindowPtr window;
+    WindowPtr root_window = NULL;
+    InputInfoPtr pInfo = grail->priv;
+    EvdevPtr pEvdev = pInfo->private;
+    int j;
+    int found_match = 0;
+    int num_clients = 0;
+    int type;
+
+    if (max_clients <= 0)
+        return 0;
+
+    if (pEvdev->flags & (EVDEV_RELATIVE_EVENTS | EVDEV_TOUCHPAD))
+    {
+        DeviceIntPtr master = pInfo->dev->u.master;
+        struct grail_coord cursor_coord;
+
+        /* If this mouse isn't hooked up to a cursor, don't do anything */
+        if (!master)
+            return 0;
+
+        cursor_coord.x = master->last.valuators[0];
+        cursor_coord.y = master->last.valuators[1];
+
+        child_window = GetWindowForGestures(grail, &cursor_coord, 1);
+    } else
+        child_window = GetWindowForGestures(grail, contacts, num_contacts);
+
+    if (!child_window)
+        return 0;
+
+    memset(clients, 0, sizeof(struct grail_client_info) * max_clients);
+
+    /*
+     * Traverse the window hierarchy looking for a window with a client
+     * selecting for one of the recognized gestures.
+     *
+     * All clients of the top most window with a match will receive events if
+     * they have selected for gestures that have been recognized, even if they
+     * have selected for different gestures between them.
+     *
+     * Once any gesture is matched on a window, propagation through the window
+     * hierarchy ends.
+     */
+    for (window = child_window; window && !found_match; window = window->parent)
+    {
+        /* No client selected for gestures on this window */
+        if (!wGestureMasks(window))
+            continue;
+
+        /* For each recognized gesture */
+        grail_mask_foreach(type, types, type_bytes)
+        {
+            /* Check if any client selected for this gesture on the window */
+            if (BitIsOn(wGestureMasks(window)->mask, type))
+            {
+                GestureClientsPtr client;
+
+                if (!root_window)
+                    for (root_window = window; root_window->parent;
+                         root_window = root_window->parent);
+
+                /* For each client that selected for gestures on this window */
+                for (client = wGestureMasks(window)->clients; client;
+                     client = client->next)
+                {
+                     /*
+                      * Check if this client selected for this gesture. Request
+                      * may be for this device or all devices.
+                      */
+                     if (BitIsOn(client->gestureMask[pInfo->dev->id], type) ||
+                         BitIsOn(client->gestureMask[0], type))
+                     {
+                         /*
+                          * Find this client in the clients array passed back to
+                          * the caller.
+                          */
+                         for (j = 0; j < num_clients; j++)
+                             if (clients[j].id.client ==
+                                 CLIENT_ID(client->resource))
+                                 break;
+
+                         /* Check if the client exists in the array yet */
+                         if (j >= num_clients)
+                         {
+                             /* We ran out of room in the array, return error */
+                             if (num_clients >= max_clients)
+                                 return -1;
+                             /* Set up new client in array */
+                             clients[j].id.client = CLIENT_ID(client->resource);
+                             clients[j].id.root = root_window->drawable.id;
+                             clients[j].id.child = child_window->drawable.id;
+                             clients[j].id.event = window->drawable.id;
+                             num_clients++;
+                         }
+
+                         /* Set this gesture bit in the client's gesture mask */
+                         SetBit(clients[j].mask, type);
+                     }
+                }
+
+                /* A match has been found, stop propagating */
+                found_match = 1;
+            }
+        }
+    }
+
+    return num_clients;
+}
+
+static void GrailEvent(struct grail *grail, const struct input_event *ev)
+{
+    InputInfoPtr pInfo = (InputInfoPtr)grail->priv;
+    EvdevProcessEvent(pInfo, (struct input_event *)ev);
+}
+
+static void GrailGesture(struct grail *grail, const struct grail_event *ev)
+{
+    InputInfoPtr pInfo = grail->priv;
+    EvdevPtr pEvdev = pInfo->private;
+    GestureEvent event;
+    float props[DIM_GRAIL_PROP];
+    int len = sizeof(GestureEvent) + ev->nprop * sizeof(grail_prop_t);
+
+    /* Client could have disappeared by now */
+    if (!clients[ev->client_id.client])
+        return;
+
+    event.type = GenericEvent;
+    event.extension = GestureReqCode;
+    event.sequenceNumber = clients[ev->client_id.client]->sequence;
+    event.evtype = 0;
+    event.time = GetTimeInMillis();
+    event.length = (len - sizeof(xEvent)) / 4;
+    event.gesture_id = ev->id;
+    event.gesture_type = ev->type;
+    event.device_id = pInfo->dev->id;
+    event.root = ev->client_id.root;
+    event.event = ev->client_id.event;
+    event.child = ev->client_id.child;
+    event.status = ev->status;
+    event.num_props = ev->nprop;
+
+    /* Override event focus point if this is a relative device */
+    if (pEvdev->flags & (EVDEV_RELATIVE_EVENTS | EVDEV_TOUCHPAD))
+    {
+        DeviceIntPtr master = pInfo->dev->u.master;
+
+        /* If this mouse isn't hooked up to a cursor, don't do anything */
+        if (!master)
+            return;
+
+        /* Note: Master device valuators are in screen coordinates */
+        event.focus_x = master->last.valuators[0];
+        event.focus_y = master->last.valuators[1];
+    } else {
+        event.focus_x = ev->pos.x;
+        event.focus_y = ev->pos.y;
+    }
+
+    memcpy(props, ev->prop, ev->nprop * sizeof(grail_prop_t));
+
+    WriteToClient(clients[ev->client_id.client], sizeof(GestureEvent), &event);
+    WriteToClient(clients[ev->client_id.client],
+                  sizeof(float) * event.num_props, props);
+}
+
+int
+GrailOpen(InputInfoPtr pInfo)
+{
+    EvdevPtr pEvdev = pInfo->private;
+
+    pEvdev->grail = malloc(sizeof(struct grail));
+    if (!pEvdev->grail) {
+        xf86Msg(X_ERROR, "evdev-grail: failed to allocate grail structure\n");
+        return -1;
+    }
+
+    memset(pEvdev->grail, 0, sizeof(struct grail));
+    pEvdev->grail->get_clients = GetClients;
+    pEvdev->grail->event = GrailEvent;
+    pEvdev->grail->gesture = GrailGesture;
+    pEvdev->grail->priv = pInfo;
+
+    if (grail_open(pEvdev->grail, pInfo->fd)) {
+        xf86Msg(X_INFO,
+                "evdev-grail: failed to open grail, no gesture support\n");
+        free(pEvdev->grail);
+        pEvdev->grail = NULL;
+        return -1;
+    }
+
+    {
+        struct grail_coord min;
+        struct grail_coord max;
+
+        min.x = screenInfo.screens[0]->x;
+        min.y = screenInfo.screens[0]->y;
+        max.x = min.x + screenInfo.screens[0]->width;
+        max.y = min.y + screenInfo.screens[0]->height;
+
+        grail_set_bbox(pEvdev->grail, &min, &max);
+    }
+
+    return 0;
+}
+
+void
+GrailClose(InputInfoPtr pInfo)
+{
+    EvdevPtr pEvdev = pInfo->private;
+
+    if (pEvdev->grail) {
+        grail_close(pEvdev->grail, pInfo->fd);
+        free(pEvdev->grail);
+        pEvdev->grail = NULL;
+    }
+}
Index: xserver-xorg-input-evdev/src/evdev-grail.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ xserver-xorg-input-evdev/src/evdev-grail.h	2011-01-28 12:20:13.601087001 +1100
@@ -0,0 +1,9 @@
+#ifndef EVDEV_GRAIL_H_
+#define EVDEV_GRAIL_H_
+
+#include <grail.h>
+
+extern int GrailOpen(InputInfoPtr pInfo);
+extern void GrailClose(InputInfoPtr pInfo);
+
+#endif /* _EVDEV_GRAIL_H_ */
Index: xserver-xorg-input-evdev/src/evdev.c
===================================================================
--- xserver-xorg-input-evdev.orig/src/evdev.c	2011-01-28 12:20:11.361087001 +1100
+++ xserver-xorg-input-evdev/src/evdev.c	2011-01-28 12:22:16.411087001 +1100
@@ -71,6 +71,13 @@
 #define MAXDEVICES MAX_DEVICES
 #endif
 
+#if USE_GRAIL
+#include "evdev-grail.h"
+#else
+#define GrailOpen(pInfo)
+#define GrailClose(pInfo)
+#endif
+
 #define ArrayLength(a) (sizeof(a) / (sizeof((a)[0])))
 
 #define MIN_KEYCODE 8
@@ -829,7 +836,7 @@
  * Process the events from the device; nothing is actually posted to the server
  * until an EV_SYN event is received.
  */
-static void
+void
 EvdevProcessEvent(InputInfoPtr pInfo, struct input_event *ev)
 {
     switch (ev->type) {
@@ -858,12 +865,20 @@
 static void
 EvdevReadInput(InputInfoPtr pInfo)
 {
+#if USE_GRAIL
+    EvdevPtr pEvdev = pInfo->private;
+#endif
     struct input_event ev[NUM_EVENTS];
     int i, len = sizeof(ev);
 
     while (len == sizeof(ev))
     {
-        len = read(pInfo->fd, &ev, sizeof(ev));
+#if USE_GRAIL
+        if (pEvdev->grail)
+            len = grail_pull(pEvdev->grail, pInfo->fd);
+        else
+#endif
+            len = read(pInfo->fd, &ev, sizeof(ev));
         if (len <= 0)
         {
             if (errno == ENODEV) /* May happen after resume */
@@ -881,6 +896,11 @@
             break;
         }
 
+#if USE_GRAIL
+        if (pEvdev->grail)
+		return;
+#endif
+
         /* The kernel promises that we always only read a complete
          * event, so len != sizeof ev is an error. */
         if (len % sizeof(ev[0])) {
@@ -1305,7 +1325,7 @@
     memset(pEvdev->old_vals, -1, num_axes * sizeof(int));
     atoms = malloc(pEvdev->num_vals * sizeof(Atom));
 
-    for (axis = ABS_X; i < MAX_VALUATORS && axis <= ABS_MAX; axis++) {
+    for (axis = ABS_X; i < MAX_VALUATORS && axis <= ABS_MISC; axis++) {
         pEvdev->axis_map[axis] = -1;
         if (!TestBit(axis, pEvdev->abs_bitmask))
             continue;
@@ -1328,7 +1348,7 @@
         return !Success;
     }
 
-    for (axis = ABS_X; axis <= ABS_MAX; axis++) {
+    for (axis = ABS_X; axis <= ABS_MISC; axis++) {
         int axnum = pEvdev->axis_map[axis];
         int resolution = 10000;
 
@@ -1722,7 +1742,6 @@
     return Success;
 }
 
-
 static int
 EvdevProc(DeviceIntPtr device, int what)
 {
@@ -1761,6 +1780,7 @@
             close(pInfo->fd);
             pInfo->fd = -1;
         }
+        GrailClose(pInfo);
         EvdevRemoveDevice(pInfo);
         pEvdev->min_maj = 0;
 	break;
@@ -1998,6 +2018,8 @@
         xf86Msg(X_PROBED, "%s: Found absolute axes\n", pInfo->name);
         pEvdev->flags |= EVDEV_ABSOLUTE_EVENTS;
 
+	GrailOpen(pInfo);
+
         if ((TestBit(ABS_X, pEvdev->abs_bitmask) &&
              TestBit(ABS_Y, pEvdev->abs_bitmask))) {
             xf86Msg(X_PROBED, "%s: Found x and y absolute axes\n", pInfo->name);
Index: xserver-xorg-input-evdev/src/evdev.h
===================================================================
--- xserver-xorg-input-evdev.orig/src/evdev.h	2011-01-28 12:20:11.351087001 +1100
+++ xserver-xorg-input-evdev/src/evdev.h	2011-01-28 12:20:13.601087001 +1100
@@ -196,8 +196,14 @@
     /* Event queue used to defer keyboard/button events until EV_SYN time. */
     int                     num_queue;
     EventQueueRec           queue[EVDEV_MAXQUEUE];
+
+#if USE_GRAIL
+    struct grail *grail;
+#endif
 } EvdevRec, *EvdevPtr;
 
+void EvdevProcessEvent(InputInfoPtr pInfo, struct input_event *ev);
+
 /* Event posting functions */
 void EvdevQueueKbdEvent(InputInfoPtr pInfo, struct input_event *ev, int value);
 void EvdevQueueButtonEvent(InputInfoPtr pInfo, int button, int value);
Index: xserver-xorg-input-evdev/src/gestureproto.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ xserver-xorg-input-evdev/src/gestureproto.h	2011-01-28 12:20:13.601087001 +1100
@@ -0,0 +1,132 @@
+/*
+ * Copyright © 2010 Canonical, Ltd.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors: Chase Douglas <chase.douglas@canonical.com>
+ *
+ */
+
+#ifndef _GESTUREPROTO_H_
+#define _GESTUREPROTO_H_
+
+#include <stdint.h>
+#include <X11/X.h>
+
+#define Window  uint32_t
+#define Time    uint32_t
+
+#define X_GestureQueryVersion                1
+#define X_GestureSelectEvents                2
+#define X_GestureGetSelectedEvents           3
+
+#define GESTUREREQUESTS (X_GestureGetSelectedEvents - X_GestureQueryVersion + 1)
+
+#define GestureAllDevices 0
+
+/**
+ * Used to select for events on a given window.
+ * Struct is followed by (mask_len * CARD8), with each bit set representing
+ * the event mask for the given type. A mask bit represents an event type if
+ * (mask == (1 << type)).
+ */
+typedef struct {
+    uint16_t    device_id;       /**< Device id to select for        */
+    uint16_t    mask_len;        /**< Length of mask in 4 byte units */
+} GestureEventMask;
+
+typedef struct {
+    uint8_t     reqType;                /**< Gesture extension major code */
+    uint8_t     ReqType;                /**< Always ::X_GestureQueryVersion */
+    uint16_t    length;                 /**< Length in 4 byte units */
+    uint16_t    major_version;
+    uint16_t    minor_version;
+} GestureQueryVersionReq;
+
+typedef struct {
+    uint8_t     repType;                /**< ::X_Reply */
+    uint8_t     RepType;                /**< Always ::X_GestureQueryVersion */
+    uint16_t    sequenceNumber;
+    uint32_t    length;
+    uint16_t    major_version;
+    uint16_t    minor_version;
+    uint32_t    pad1;
+    uint32_t    pad2;
+    uint32_t    pad3;
+    uint32_t    pad4;
+    uint32_t    pad5;
+} GestureQueryVersionReply;
+
+typedef struct {
+    uint8_t     reqType;                /**< Gesture extension major code */
+    uint8_t     ReqType;                /**< Always ::X_GestureSelectEvents */
+    uint16_t    length;                 /**< Length in 4 byte units */
+    Window      window;
+    GestureEventMask mask;
+} GestureSelectEventsReq;
+
+typedef struct {
+    uint8_t     reqType;                /**< Gesture extension major code */
+    uint8_t     ReqType;                /**< Always ::X_GestureGetSelectedEvents */
+    uint16_t    length;                 /**< Length in 4 byte units */
+    Window      window;
+} GestureGetSelectedEventsReq;
+
+typedef struct {
+    uint8_t     repType;                /**< Gesture extension major opcode */
+    uint8_t     RepType;                /**< Always ::X_GestureGetSelectedEvents */
+    uint16_t    sequenceNumber;
+    uint32_t    length;
+    uint16_t    num_masks;              /**< Number of GestureEventMask structs
+                                             trailing the reply */
+    uint16_t    pad0;
+    uint32_t    pad1;
+    uint32_t    pad2;
+    uint32_t    pad3;
+    uint32_t    pad4;
+    uint32_t    pad5;
+} GestureGetSelectedEventsReply;
+
+typedef struct
+{
+    uint8_t     type;                   /**< Always GenericEvent */
+    uint8_t     extension;              /**< Gesture extension offset */
+    uint16_t    sequenceNumber;         /**< Xevent sequence number */
+    uint32_t    length;                 /**< Length in 4 byte uints */
+    uint16_t    evtype;                 /**< X generic event type */
+    uint16_t    gesture_id;             /**< Unique ID for gesture */
+    uint16_t    gesture_type;           /**< Gesture type (zoom, rotate, etc.) */
+    uint16_t    device_id;              /**< Device that generated this gesture */
+    Time        time;                   /**< Time of gesture event */
+    Window      root;                   /**< Root window event occurred on */
+    Window      event;                  /**< Window selecting this event for a client */
+    Window      child;                  /**< Top-most window of gesture event */
+/* └──────── 32 byte boundary ────────┘ */
+    float       focus_x;                /**< Always window coords, 16.16 fixed point */
+    float       focus_y;                /**< Relative to event window */
+    uint16_t    status;                 /**< Gesture event status */
+    uint16_t    num_props;               /**< Number of properties for gesture event */
+/* └──── Gesture properties below ────┘ */
+} GestureEvent;
+
+#undef Window
+#undef Time
+
+#endif /* _GESTUREPROTO_H_ */
