http://bugs.freedesktop.org/show_bug.cgi?id=18150

Index: xserver-xorg-input-evdev/src/evdev.c
===================================================================
--- xserver-xorg-input-evdev.orig/src/evdev.c	2008-10-21 18:20:00.000000000 +0300
+++ xserver-xorg-input-evdev/src/evdev.c	2008-10-21 18:20:22.000000000 +0300
@@ -410,7 +410,7 @@
     }
 }
 
-#define TestBit(bit, array) (array[(bit) / LONG_BITS]) & (1 << ((bit) % LONG_BITS))
+#define TestBit(bit, array) (array[(bit) / LONG_BITS]) & (1l << ((bit) % LONG_BITS))
 
 static void
 EvdevPtrCtrlProc(DeviceIntPtr device, PtrCtrl *ctrl)
