diff -Nurp patched/src/evdev.c working/src/evdev.c
--- patched/src/evdev.c	2009-02-10 09:31:26.000000000 -0800
+++ working/src/evdev.c	2009-02-10 09:43:06.000000000 -0800
@@ -1214,8 +1214,10 @@ EvdevCacheCompare(InputInfoPtr pInfo, BO
         goto error;
     }
 
-    if (compare && strcmp(pEvdev->name, name))
+    if (compare && strcmp(pEvdev->name, name)) {
+        xf86Msg(X_ERROR, "device name changed: %s != %s\n", pEvdev->name, name);
         goto error;
+    }
 
     if (ioctl(pInfo->fd,
               EVIOCGBIT(0, sizeof(bitmask)), bitmask) < 0) {
@@ -1223,8 +1225,10 @@ EvdevCacheCompare(InputInfoPtr pInfo, BO
         goto error;
     }
 
-    if (compare && memcmp(pEvdev->bitmask, bitmask, sizeof(bitmask)))
+    if (compare && memcmp(pEvdev->bitmask, bitmask, sizeof(bitmask))) {
+        xf86Msg(X_ERROR, "device bitmask has changed\n");
         goto error;
+    }
 
 
     if (ioctl(pInfo->fd,
@@ -1233,8 +1237,10 @@ EvdevCacheCompare(InputInfoPtr pInfo, BO
         goto error;
     }
 
-    if (compare && memcmp(pEvdev->rel_bitmask, rel_bitmask, sizeof(rel_bitmask)))
+    if (compare && memcmp(pEvdev->rel_bitmask, rel_bitmask, sizeof(rel_bitmask))) {
+        xf86Msg(X_ERROR, "device rel_bitmask has changed\n");
         goto error;
+    }
 
     if (ioctl(pInfo->fd,
               EVIOCGBIT(EV_ABS, sizeof(abs_bitmask)), abs_bitmask) < 0) {
@@ -1242,8 +1248,10 @@ EvdevCacheCompare(InputInfoPtr pInfo, BO
         goto error;
     }
 
-    if (compare && memcmp(pEvdev->abs_bitmask, abs_bitmask, sizeof(abs_bitmask)))
+    if (compare && memcmp(pEvdev->abs_bitmask, abs_bitmask, sizeof(abs_bitmask))) {
+        xf86Msg(X_ERROR, "device abs_bitmask has changed\n");
         goto error;
+    }
 
     if (ioctl(pInfo->fd,
               EVIOCGBIT(EV_KEY, sizeof(key_bitmask)), key_bitmask) < 0) {
@@ -1251,8 +1259,10 @@ EvdevCacheCompare(InputInfoPtr pInfo, BO
         goto error;
     }
 
-    if (compare && memcmp(pEvdev->key_bitmask, key_bitmask, sizeof(key_bitmask)))
+    if (compare && memcmp(pEvdev->key_bitmask, key_bitmask, sizeof(key_bitmask))) {
+        xf86Msg(X_ERROR, "device key_bitmask has changed\n");
         goto error;
+    }
 
     if (ioctl(pInfo->fd,
               EVIOCGBIT(EV_LED, sizeof(led_bitmask)), led_bitmask) < 0) {
@@ -1260,8 +1270,10 @@ EvdevCacheCompare(InputInfoPtr pInfo, BO
         goto error;
     }
 
-    if (compare && memcmp(pEvdev->led_bitmask, led_bitmask, sizeof(led_bitmask)))
+    if (compare && memcmp(pEvdev->led_bitmask, led_bitmask, sizeof(led_bitmask))) {
+        xf86Msg(X_ERROR, "device led_bitmask has changed\n");
         goto error;
+    }
 
     memset(absinfo, 0, sizeof(absinfo));
 
@@ -1277,8 +1289,10 @@ EvdevCacheCompare(InputInfoPtr pInfo, BO
         }
     }
 
-    if (compare && memcmp(pEvdev->absinfo, absinfo, sizeof(absinfo)))
-            goto error;
+    if (compare && memcmp(pEvdev->absinfo, absinfo, sizeof(absinfo))) {
+        xf86Msg(X_ERROR, "device absinfo has changed\n");
+        goto error;
+    }
 
     /* cache info */
     if (!compare)
